tasks_week1_Gal_Malka/src/Vector.cpp

```diff
@@ -0,0 +1,355 @@
+
+/*
+* operator= - copy the initializer list data into the vector and deallocate previous data.
+*/
+Vector& Vector::operator= (const std::initializer_list<int> initializer) {
+	_size = initializer.size();
+	_capacity = _size;
+	delete[] _data;
+	_data = new int[_capacity];
+	copy_data(initializer.begin(), &_data, _size);
+	return *this;
+};
+
+/*
+* swap - swaps the data between two vectors.
+*/
+void Vector::swap(Vector& other) {
+
+	//swap capacity
+	other._capacity += _capacity;
+	_capacity = other._capacity - _capacity ;
+	other._capacity -= _capacity;
+
+	//swap size
+	other._size += _size;
+	_size = other._size - _size;
+	other._size -= _size;
+
+	//swap data
+	int* temp = other._data;
+	other._data = _data;
+	_data = temp;
+
+	temp = nullptr;
+}
+
+/*
+* push_back - append a value to the end of the vector. if the vector's capacity is not enaugh,
+* than it will allocate more memory.
+*/
+void Vector::push_back(const int& value) {
+
+	if (_size == _capacity) {
+
+		if (_capacity == 0) {
+
+			_capacity = 2;
+			_data = new int[_capacity];
+		}
+
+		else {
+			//allocate more memory and copy data.
+			_capacity = _capacity < 128 ? _capacity * 2 : _capacity * 1.5;
+			int* newData = new int[_capacity];
+			copy_data(_data, &newData, _size);
+			delete[] _data;
+			_data = newData;
+			newData = nullptr;
+		}
+	}
+
+	*(_data +_size) = value;
+	_size++;
+
+}	
+
+/*
+* clear - destroy all data
+*/
+void Vector::clear() noexcept {
+
+	_size = 0;
+	_capacity = 0;
+	if (_data)
+		delete[] _data;
+	_data = nullptr;
+}
+
+/*
+* resize - change the size of the vector to newSize
+*/
+void Vector::resize(size_t newSize) {
+
+	if (newSize > _capacity) {
+		//allocate more memory and copy data.
+		_capacity = newSize < 128 ? newSize * 2 : newSize * 1.5;
+		int* temp = _data;
+		_data = new int[_capacity];
+		copy_data(_data, &temp, _size);
+		delete[] temp;
+	}
+
+	//initialize the remaining values to 0.
+	for (; _size < newSize; _size++) {
+		_data[_size] = 0;
+	}
+
+	_size = newSize;
+}
+
+void Vector::resize(size_t newSize, const int& value = 0) {
+
+	if (newSize > _capacity) {
+		//allocate more memory and copy data.
+		_capacity = newSize < 128 ? newSize * 2 : newSize * 1.5;
+		int* temp = _data;
+		_data = new int[_capacity];
+		copy_data(_data, &temp, _size);
+		delete[] temp;
+	}
+	//initialize  the remaining values to value.
+	for (; _size < newSize; _size++) {
+		_data[_size] = value;
+	}
+}
+
+// copy_data - a private function that copy vectors data to another vector data.
+void Vector::copy_data(const int* a, int** b,const size_t& size) {
+	
+	for (size_t i = 0; i < size; i++) {
+		*(*b + i) = *(a + i);
+	}
+}
+
+/*
+* operator== - compares std::vector of integers to Vector object to compare results for tests purposes.
+*/
+bool Vector::operator==(const std::vector<int>& other) const {
+
+	//compare sizes
+	if (_size != other.size()) return false;
+
+	//compare values
+	for (size_t i = 0; i < _size ; i++) {
+		if (_data[i] != other[i]) {

```

Comment:
It's a waste to use `other`'s `operator[]`, since every call performs the limits check (the if at the beginning). You can just use `other.data()` directly for better performance